Add stats API

steveburkett opened the pull request:
Add a new stats API for fetching book index status.

Also, remove the unused prototype controller.

https://app.zenhub.com/workspaces/openstax-unified-5b71aabe3815ff014b102258/issues/openstax/unified/201

![image](https://user-images.githubusercontent.com/352161/61489009-7d8e2b80-a95e-11e9-84f8-f7182cb5628c.png)

steveburkett committed:
Add stats API

Add a new stats API for fetching book index status.

Also, remove the unused prototype controller.

steveburkett committed:
Add specs to stats API

jpslav reviewed (changes requested):
Thanks @steveburkett !

steveburkett commented on app/controllers/api/v0/stats_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  def stats
+    started_at = Time.now
+
+    stats_results = Stats.new.call
```
@jpslav it's not good practice for constructors to do work.  they're to set up the class invariant.  hence the use of call (to do the work, as a service object would).  #call could be cleared before use to set it up a second time.

steveburkett commented on app/controllers/api/v0/stats_controller.rb:
```diff
@@ -0,0 +1,30 @@
+class Api::V0::StatsController < Api::V0::BaseController
+
+  swagger_path '/stats' do
```
well..."num_docs" isnt really a status of the index is it?  that feels more like stats for the index.  and but "state" is certainly more of a status.  will change to status tho....

steveburkett replied:
sounds like /info won the vote

steveburkett commented on app/controllers/api/v0/stats_controller.rb:
```diff
@@ -0,0 +1,30 @@
+
+    response = Api::V0::Bindings::StatResults.new.build_from_hash(stats_results.with_indifferent_access)
+
+    response.overall_took_ms = ((Time.now - started_at)*1000).round
```
personally...i think we should use this as a standard practice.   a top level meta layer on all api calls that gives time_spent and other meta information at the top level.  if that shows as really long...then it's a easy indication that's somethings wrong w/ accessing dynamo or elasticsearch.  

that's my vote...but of course, can remove it if you want :)

steveburkett commented on app/services/stats.rb:
```diff
@@ -0,0 +1,67 @@
+        update_stat(index: es_index.first,
+                    value_sym: :num_docs,
+                    value: es_index.second["primaries"]["docs"]["count"])
+      end
```
also..stuff like this would be "stats", not "status" eh? :)

steveburkett commented on app/services/stats.rb:
```diff
@@ -0,0 +1,67 @@
+  def load_dynamo_books
+    books = BookIndexState.all
+    books.each do |book|
+      update_stat(index: "#{book.book_version_id}_#{book.indexing_strategy_name}".downcase,
```
yup!  we already had this method on Books::Index

steveburkett committed:
Code review fixes

Rename endpoint to /info
Add created_at to returned info
Fix tests
Re-record VCR cassettes

jpslav commented on app/services/index_info.rb:
```diff
@@ -0,0 +1,72 @@
+  def es_created_at(index_name)
+    index = OsElasticsearchClient.instance.indices.get(index: index_name)
+    es_created_at_ms = index[index_name]["settings"]["index"]["creation_date"]
+    DateTime.strptime((es_created_at_ms.to_i/1000).to_s,'%s')
```
Let's use UTC and format as ISO8601.  

```ruby
Time.at(es_created_at_ms.to_i/1000).utc.iso8601
```

which gives e.g. `"2019-07-02T20:56:20Z"`

We tend to use iso 8601 in various apps, and UTC is nice to not get into time zone issues.  

Also when we list the indexes we can sort on this using string comparison.

May have missed it, but can you please make sure there's a test for this timestamp code?

jpslav commented on app/controllers/api/v0/swagger/models/info.rb:
```diff
@@ -14,14 +14,19 @@ module Api::V0::Swagger::Models::Stat
+    property :created_at do
+      key :type, :string
+      key :readOnly, true
+      key :description, "The (elasticsearch) created_at time"
```
add that it is given in ISO 8601

jpslav commented on app/services/index_info.rb:
```diff
@@ -0,0 +1,72 @@
+  def results
+    {
+      es_version: es_version,
+      book_indexes: all_the_books
```
Somewhere can we sort the book indexes by creation time (or something).  I'd prefer to know that the results come back in a consistent order on every call to `/info` and I'm unsure if we otherwise have that guarantee.

jpslav commented on app/services/index_info.rb:
```diff
@@ -0,0 +1,72 @@
+  end
+
+  def es_created_at(index_name)
+    index = OsElasticsearchClient.instance.indices.get(index: index_name)
```
Is it expensive to call `instance.indices` over and over?

steveburkett committed:
Round 2 of code review fixes

Change created_at time to ISO8601 format
Memoize ElasticSearch indices call
Sort books return by created_at, most recent first.
